pennylane/transforms/optimization/pattern_matching.py

```diff
@@ -0,0 +1,1594 @@
+        apply(op)
+
+    # After optimization, simply apply the measurements
+    for obs in observables:
+        obs._wires = Wires([inverse_wires_map[wire] for wire in obs.wires.tolist()])
+
+    for m in measurements:
+        apply(m)
+
+
+def pattern_matching(circuit_dag, pattern_dag):
+    r"""Function that applies the pattern matching algorithm and returns the list of maximal matches.
+
+    Args:
+        circuit_dag (.CommutationDAG): A commutation DAG representing the circuit to be optimized.
+        pattern_dag(.CommutationDAG): A commutation DAG representing the pattern.
+
+    Returns:
+        list(Match): the list of maximal matches.
+    """
+    # Match list
+    match_list = []
+
+    # Loop through all possible initial matches
+    for node_c in circuit_dag.get_nodes():
+        for node_p in pattern_dag.get_nodes():
+            # Initial matches between two identical gates (No qubits comparison)
+            if _compare_operation_without_qubits(node_c[1], node_p[1]):
+
+                circuit_range = range(0, circuit_dag.num_wires)
+
+                # Fix qubits from the first (target fixed and control restrained)
+                not_fixed_qubits_confs = _not_fixed_qubits(
+                    circuit_range, node_c[1].wires, pattern_dag.num_wires - len(node_p[1].wires)
+                )
+
+                # Loop over all possible qubits configurations given the first match constrains
+                for not_fixed_qubits_conf in not_fixed_qubits_confs:
+                    for not_fixed_qubits_conf_permuted in itertools.permutations(
+                            not_fixed_qubits_conf
+                    ):
+                        not_fixed_qubits_conf_permuted = list(not_fixed_qubits_conf_permuted)
+                        for first_match_qubits_conf in _first_match_qubits(
+                                node_c[1], node_p[1], pattern_dag.num_wires
+                        ):
+                            # Qubits mapping between circuit and pattern
+                            qubits_conf = _merge_first_match_and_permutation(
+                                first_match_qubits_conf, not_fixed_qubits_conf_permuted
+                            )
+
+                            # Update wires, target_wires, control_wires
+                            wires, target_wires, control_wires = _update_qubits(
+                                circuit_dag, qubits_conf
+                            )
+
+                            # Forward match part of the algorithm
+                            forward = ForwardMatch(
+                                circuit_dag,
+                                pattern_dag,
+                                node_c[0],
+                                node_p[0],
+                                wires,
+                                target_wires,
+                                control_wires,
+                            )
+                            forward.run_forward_match()
```

Comment:
Does it make sense to use a class here and with `BackwardMatch`? Can we get the same job done with a function? The function could still create some kind of data class internally, but I think this logic could be encapsulated by a function and be more readable and maintainable.